[GPU Process] Vertical text is incorrectly displaced

litherum opened the pull request:
#### d6a26214265522f7390962f6f4e88d9ba368f46d
<pre>
[GPU Process] Vertical text is incorrectly displaced
<a href="https://bugs.webkit.org/show_bug.cgi?id=239484">https://bugs.webkit.org/show_bug.cgi?id=239484</a>
&lt;rdar://91187123&gt;

Reviewed by Said Abou-Hallawa and Darin Adler.

Implement computeVerticalAdvancesFromPositions() as the mathematical inverse of
fillVectorWithVerticalGlyphPositions().

Covered by existing tests (with the GPU process enabled).
imported/w3c/web-platform-tests/css/css-writing-modes/full-width-003.html has always
failed, but it was reported as passing because it&apos;s an -expected-mismatch test and
there was floating point precision differences. This patch correctly marks the test
as failing.

* Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
(WebCore::computeHorizontalAdvancesFromPositions):
(WebCore::computeVerticalAdvancesFromPositions):
(WebCore::DrawGlyphsRecorder::recordDrawGlyphs):
(WebCore::computeAdvancesFromPositions): Deleted.
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::computeBaseOverallTextMatrix):
(WebCore::computeBaseVerticalTextMatrix):
(WebCore::fillVectorWithVerticalGlyphPositions):

Canonical link: <a href="https://commits.webkit.org/251981@main">https://commits.webkit.org/251981@main</a>
</pre>

darinadler commented on Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
```diff
@@ -237,16 +237,56 @@ void DrawGlyphsRecorder::recordEndLayer(CGRenderingStateRef, CGGStateRef gstate)
+
+    auto syntheticObliqueLessTextMatrix = computeBaseVerticalTextMatrix(computeBaseOverallTextMatrix(std::nullopt)); // See fillVectorWithVerticalGlyphPositions(), which describes what this is.
+
+    auto transformPoint = [&](const CGPoint& position, const CGSize& translation) -> CGPoint {
```
Not sure that taking these by reference instead of value is good for performance. Could use just `CGPoint` and `CGSize` instead of `const&`.

litherum replied:
Happy to do that.

What's the rule here? I thought the rule was "is it bigger than a register?"

darinadler commented on Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
```diff
@@ -298,21 +338,17 @@ void DrawGlyphsRecorder::recordDrawGlyphs(CGRenderingStateRef, CGGStateRef gstat
     auto textMatrix = m_originalTextMatrix;
-    auto initialPenPosition = textMatrix.mapPoint(positions[0]);
 
+    AdvancesAndInitialPosition advancesAndInitialPosition;
```
Would be nice to have a shorter name for this local variable. Even naming it just `advances` would probably make the code more readable even though the name is less precise.

darinadler replied:
Or naming it `position`.

darinadler commented on Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
```diff
@@ -263,15 +263,17 @@ static void fillVectorWithVerticalGlyphPositions(Vector<CGPoint, 256>& positions
     // is at the point labeled B in the above diagram, in user coordinates.
     auto position = CGPointMake(point.x(), point.y() + ascentDelta);
 
+    static const auto constantSyntheticObliqueLessTextMatrix = computeBaseVerticalTextMatrix(computeBaseOverallTextMatrix(std::nullopt));
```
Consider `constantSyntheticTextMatrixOmittingOblique`?

darinadler commented on Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
```diff
@@ -237,16 +237,56 @@ void DrawGlyphsRecorder::recordEndLayer(CGRenderingStateRef, CGGStateRef gstate)
+
+    ASSERT(count); // Because we say "positions[0]" below.
+
+    auto syntheticObliqueLessTextMatrix = computeBaseVerticalTextMatrix(computeBaseOverallTextMatrix(std::nullopt)); // See fillVectorWithVerticalGlyphPositions(), which describes what this is.
```
Not super fond of "oblique less" as a way to write "without oblique". No sure that "obliqueless" as a single word would be better. Maybe, "omitting oblique", so, `syntheticTextMatrixOmittingOblique`?

darinadler reviewed (approved):
I didn’t spot any problems and I think this takes all of Said’s suggestions. I have a few minor coding style tweak thoughts that are entirely optional.

litherum committed:
[GPU Process] Vertical text is incorrectly displaced
https://bugs.webkit.org/show_bug.cgi?id=239484
<rdar://91187123>

Reviewed by Said Abou-Hallawa and Darin Adler.

Implement computeVerticalAdvancesFromPositions() as the mathematical inverse of
fillVectorWithVerticalGlyphPositions().

Covered by existing tests (with the GPU process enabled).
imported/w3c/web-platform-tests/css/css-writing-modes/full-width-003.html has always
failed, but it was reported as passing because it's an -expected-mismatch test and
there was floating point precision differences. This patch correctly marks the test
as failing.

* Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
(WebCore::computeHorizontalAdvancesFromPositions):
(WebCore::computeVerticalAdvancesFromPositions):
(WebCore::DrawGlyphsRecorder::recordDrawGlyphs):
(WebCore::computeAdvancesFromPositions): Deleted.
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::computeBaseOverallTextMatrix):
(WebCore::computeBaseVerticalTextMatrix):
(WebCore::fillVectorWithVerticalGlyphPositions):

Canonical link: https://commits.webkit.org/251981@main
